### ISSUE TITLE
Packages cannot depend_on both blas and lapack separately

### ISSUE THREAD

**Issue by adamjstewart** [LLNL/spack#538]
I am writing a package for qrupdate, which depends on blas and lapack.
```
$ spack graph qrupdate
o  qrupdate
|\
o |  lapack
 /
o  blas
```
When I try to build or even spec this package, I see the following error message.
```
$ spack spec qrupdate
Input spec
------------------------------
  qrupdate

Normalized
------------------------------
  qrupdate
      ^blas
      ^lapack

Concretized
------------------------------
==> Error: Cannot depend on 'openblas' twice
```
The obvious solution is to depend_on openblas instead of blas and lapack. I'm just wondering if this behavior is expected or not?

**Comment by tgamblin** [comment]
This is an error in concretization -- try forcing the lapack implementation to something other than OpenBLAS, which provides both.  I want to find some time to look over the lapack/BLAS stuff that @alalazo has recently put in, but I have not had time.  I was becoming concerned about packages that provide both BLAS and LAPACK because we care about composing with vendor BLAS implementations, but it looks like he recently updated his PR so there may be a solution there.

I'll also obviously look at the concretization error here.

**Comment by adamjstewart** [comment]
Yeah, I'm not too worried. I'll tell it to depend_on openblas for now. Just wanted to report the error.

**Comment by mplegendre** [comment]
Todd, are you reproducing this in any packages we have in the repo?  I'm suspicious that I'm the cause of any recent concretization errors, though I'm not seeing anything with other lapack packages.

**Comment by tgamblin** [comment]
I have tested the openblas case and I *believe* I added a case in the concretization tests for this.  But it would be good to make a reproducer for @adamjstewart's case here and make sure it's in the tests.

We may need to bite the bullet and start putting in smarter constraint solving.  I worry that greediness is going to catch up to us eventually.  This may be a case where concretization should've backtracked but didn't.

**Comment by deleeke** [comment]
I think I experienced a similar problem today.
(I'm using the most recent spack)

I was trying to install petsc and was getting this error:
```
me@myhost:~/Repos/spack$ spack install petsc
==> Error: Cannot depend on 'openblas' twice
```
so I did this:

```
me@myhost:~/Repos/spack$ spack edit petsc
```
And I commented out the dependency on openblas. Then when I attemped to "spack install petsc ~mpi" it proceeded without error.

**Comment by tgamblin** [comment]
@mplegendre @adamjstewart I implemented less greedy concretization in #549.  See if this works for you.

There were two issues here: 

  1. The new concretization code was not checking to see if the concrete replacement for a virtual was already in the spec.  So, if you resolve `lapack` to `openblas`, then try to resolve `bias` to `openblas`, a sanity check fails when you try to add that second redundant dependency.

  2. The order choices are made in during concretization can matter.  If you resolve the first `lapack` to `openblas`, then you end up picking `netlib-blas` for `bias`, it won't work b/c you already have a `bias` provider.  The concretizer needed a little more foresight.  Now it will test each concretization candidate until it finds one that doesn't immediately violate any constraints.  It won't backtrack more than one substitution, but I think this should handle the current level of virtual dep complexity pretty well. I do not know of any vdeps that expand to a concrete package that itself has another vdep, which is where you could potentially confuse Spack again.

Let me know if the PR works for you.

@mplegendre: I put some more detailed comments in the commits, FYI.

**Comment by tgamblin** [comment]
@deleeke: That'll work fine for `petsc` but you shouldn't have to do it :smile:.  This was a bug in some of Spack's dependency resolution code.
